[javinchua] iP

javinchua opened the pull request:
### Code Buddy

> “One of the secrets of getting more done is to make a TO-DO List every day, keep it visible, and use it as a guide to action as you go through the day.” — Jean de La Fontaine

Are you struggling with the following things:
- Keeping track of tasks
- Forgetting tasks deadlines
- Forgetting the timeline of an event

Do not fret! Code Buddy will solve these struggles for you 😄 

All you need to do is:

1. Run the .jar file
2. Key in tasks using `todo` or `event` or `deadline` commands followed by the description of the tasks
3. Use `mark 2` to mark the task as done
4. Use `delete 2` to delete the task
5. Use `list` to display the list of tasks

List of features:

- [x] Adding and removing tasks
- [ ] Improve UI

Below is a code snippet of the main method running the application.
```java
    public static void main(String[] args) {
        new Duke().run();
    }
``` 

This app is ~easy~ **_super easy_** to use. Fork the application from the [link](https://github.com/javinchua/ip/) here

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

javinchua committed:
Rename, Greet, Exit

javinchua committed:
Echo

javinchua committed:
Add, List

javinchua committed:
Mark as Done

javinchua committed:
ToDos, Events, Deadlines

javinchua committed:
Automated Text UI Testing

javinchua committed:
Handle Errors

javinchua committed:
Delete

javinchua committed:
Use Enums

javinchua committed:
Refactor print of types

Commit:
Fix index

javinchua committed:
Revert "Fix index"

This reverts commit e627e0733253df394c6ed70f48618e4ad6654cc1.

javinchua committed:
Revert "Revert "Fix index""

This reverts commit 4f714cd8480bdcd35519ed790c8a1c4bc117331d.

javinchua committed:
Save Tasks to Harddrive

Tasks that are added or deleted disappear after the chatbot is reloaded

Users will not be able to access previous tasks

Save the tasks in the hard disk automatically whenever the task list changes. Load the data from the hard disk when the chatbot starts up.

javinchua committed:
Teach chatbot how to understand dates and times

Output of datetime of task depended on user's input format

Unclear understanding of date time

Use DateTimeFormatter to format the time and present it clearer to the user

javinchua committed:
Refactor the code to extract out closely related code as classes

javinchua committed:
Organise into packages

javinchua committed:
Add gradle support

javinchua committed:
Add JUnit Tests

javinchua committed:
Add JUnit test cases

javinchua committed:
Fix printing statements

javinchua committed:
Edit gradle file

javinchua committed:
Edit JavaDocs to be concise

javinchua committed:
Follow coding standard

Tweak the code to follow Java Coding standards

javinchua committed:
Find keywords

Given a keyword, users can search for tasks containing that keyword.

javinchua committed:
Merge branch 'A-CodingStandard'

javinchua committed:
Merge branch 'branch-Level-9'

javinchua committed:
Update Parser.java

javinchua committed:
Modify code based on checkstyle violations

javinchua committed:
Implementing Java GUI

cyaoxuan commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,113 @@
+public class Storage {
+
+    /** File path to where the tasks are saved. */
+    private static final String filePath = "./data/hardrive.txt";
```
Since this is a constant, should the name be in full caps (i.e. FILE_PATH)?

cyaoxuan commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,113 @@
+    private static final String filePath = "./data/hardrive.txt";
+
+    /** Directory path to where the file is located. */
+    private static final String dataFolderPath = "./data";
```
Same as above, should the path be named in full caps?

cyaoxuan commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,113 @@
+     *
+     * @param store The list of tasks to save.
+     */
+    public void save(ArrayList<Task> store) {
```
Perhaps the name could be more noun-sounding and descriptive, such as taskList?

cyaoxuan commented on src/test/java/duke/tasks/DeadlineTest.java:
```diff
@@ -0,0 +1,33 @@
+
+    // Testing the creation of a Deadline task and its default values.
+    @Test
+    public void createDeadlineTest() {
```
Should this function just be called `createDeadline` as per the naming standard for tests as shown [here](https://se-education.org/guides/conventions/java/intermediate.html#naming:~:text=Underscores%20may%20be%20used%20in%20test%20method%20names%20using%20the%20following%20three%20part%20format%20featureUnderTest_testScenario_expectedBehavior())?

cyaoxuan commented on src/test/java/duke/util/ParserTest.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    @Test
+    public void parse_deadlineCommandWithDescriptionAndBy_addDeadlineCommandReturned() throws DukeException, EmptyTaskException, InvalidDateTimeException {
```
Consider splitting this line up as it goes beyond the 120 character limit?

cyaoxuan reviewed (commented):
It is evident you put A LOT of effort in the code, I'm extremely impressed!! Your code is really clean and well documented, and I am just nitpicking at this point. Good job! :D

timetraveller-123 commented on src/main/java/duke/Duke.java:
```diff
@@ -68,4 +72,29 @@ public void run() {
+
+
+    @FXML
+    public String getResponse(String input) {
```
Perhaps could consider adding javadocs to getResponse since it is a public method.

timetraveller-123 commented on src/main/java/duke/gui/DialogBox.java:
```diff
@@ -0,0 +1,65 @@
+        setAlignment(Pos.TOP_LEFT);
+    }
+
+    public static DialogBox getUserDialog(String text, Image img) {
```
Perhaps could consider adding javadocs to getUserDialog since it is a public method.

timetraveller-123 commented on src/main/java/duke/gui/DialogBox.java:
```diff
@@ -0,0 +1,65 @@
+        return new DialogBox(text, img);
+    }
+
+    public static DialogBox getDukeDialog(String text, Image img) {
```
Perhaps could consider adding javadocs to getDukeDialog since it is a public method.

timetraveller-123 commented on src/main/java/duke/gui/Main.java:
```diff
@@ -0,0 +1,33 @@
+    private Duke duke = new Duke();
+
+    @Override
+    public void start(Stage stage) {
```
Perhaps could consider adding javadocs to start since it is a public method.

timetraveller-123 reviewed (commented):
Overall coding standard followed everywhere except for a few missing Javadocs. Good job.

HEEaZ commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,100 @@
+            } finally {
+                ui.showLine();
+            }
+
```
Perhaps we can omit these blank newlines?

HEEaZ commented on src/main/java/duke/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,53 @@
+     */
+    @Override
+    public String execute(TaskList tasks, Ui ui, Storage storage) throws InvalidIndexException {
+        String output = tasks.unmarkTask(index);
```
Maybe we can rename 'output' to be more informative? Perhaps confimrationMessage?

HEEaZ commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,51 @@
+     *
+     * @return A {@code String} representing the deadline of the task.
+     */
+    public String getBy() {
```
Maybe we can consider finding another name for this method (and possibly instance variable)? `getBy` might not be the most intuitive to understand 😸

HEEaZ commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,117 @@
+public class Task {
+    private TaskType type;
+    private String description;
+    private boolean done;
```
Perhaps we can rename this to `isDone` to make it even clearer that the variable name is a boolean?

HEEaZ commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,153 @@
+    public String findTasks(String keyword) {
+        String output = "";
+        output += "Here are the matching tasks in your list:\n";
+        boolean found = false;
```
maybe we can name this variable as `isFound` instead of `found`?

HEEaZ reviewed (commented):
Super clean and concise code! Only comment I would hv is to perhaps align boolean variable namings to the convention in the coding standard

javinchua committed:
Modify GUI

Text gets truncated when text is longer than the height of the dialog

This results in tasks not being visible to the user

Changing the .fxml files to allow growing of dialog container

This allows users to see all tasks

javinchua committed:
Add assertions

There is no check to ensure tasklist is instantiated correctly and tasks are added and deleted correctly

We can add checks to see if the code is behaving correctly.

We add assert statements to ensure the code is behaving appropriately.

javinchua committed:
Refactor TaskList for enhanced readability and maintainability

The TaskList class has areas with redundant code and unclear string constructions. Additionally, methods like string concatenations in loops are not optimized.

To ensure the maintainability of the code, increase readability, and optimize performance, it's essential to refactor the areas of concern. Moreover, it's crucial to prevent potential bugs and side effects from external list modifications.

Refactor string building using String.format().
Centralize index validation with validateIndex().
Ensure a deep copy of the input ArrayList during object construction.
Adopt StringBuilder for methods with loop-based string concatenations.
Enhance method documentation.
Add null-checks for enhanced error handling.

Using String.format() and StringBuilder improves clarity and performance, respectively. Centralizing the validation logic prevents code repetition, making the codebase more maintainable. Deep copying ensures the TaskList object remains in a valid state, isolated from external modifications. Enhanced documentation aids in understanding for future developers. Null checks are a good defensive programming measure.

javinchua committed:
Add javadocs

javinchua committed:
Add Updating of Tasks

javinchua committed:
Merge pull request #1 from javinchua/branch-A-CodeQuality

Branch a code quality

javinchua committed:
Merge branch 'master' into branch-BCD-Extension

javinchua committed:
Merge branch 'master' into branch-A-Assertions

javinchua committed:
Merge pull request #2 from javinchua/branch-BCD-Extension

Branch bcd extension

javinchua committed:
Merge branch 'master' into branch-A-Assertions

javinchua committed:
Merge pull request #3 from javinchua/branch-A-Assertions

Branch a assertions

javinchua committed:
Fix checkstyle code

javinchua committed:
Update test case

javinchua committed:
update based on naming conventions

javinchua committed:
Update User Guide

javinchua committed:
Merge branch 'branch-A-UserGuide'

javinchua committed:
Update build.gradle

javinchua committed:
Update README.md
